### ISSUE TITLE
For LM output head FP32 precision, is the GEMM using TF32 Tensor Core or FP32 CUDA Core unit?

### ISSUE THREAD

**Issue by TheTinyTeddy** [MiniMax-AI/MiniMax-M1#3]
Many thanks for the great work!

I was wondering is the LM output head FP32 precision computation using TF32 Tensor Core or FP32 CUDA Core unit?

My second question is what are the different precision used in the forward kernel and the backward kernel?

Kind regards

**Comment by ZZBoom** [comment]
During pre-training, the lm head was used with bf16 precision, i.e., bf16 tensor core. During the RL process, especially in the generation phase, due to the larger activation values of our model, logits required higher precision. Therefore, during training, the lm head was switched to fp32 for both forward and backward passes - and did not enable TF32 throughout the training phase.

**Comment by TheTinyTeddy** [comment]
> During pre-training, the lm head was used with bf16 precision, i.e., bf16 tensor core. During the RL process, especially in the generation phase, due to the larger activation values of our model, logits required higher precision. Therefore, during training, the lm head was switched to fp32 for both forward and backward passes - and did not enable TF32 throughout the training phase.

Just to confirm, do you mean the actual GEMM computation is using purely FP32 CUDA Core SIMT kernel (i.e. it is NOT FP32 datatype A and FP32 datatype B with the actual GEMM computation using TF32 Tensor Core, since Ampere, Hopper or Blackwell GPUs only support TF32 precision Tensor Core computation and do not support FP32 precision Tensor Core computation in hardware)?
